power-sequencer/mihawk-cpld.cpp

```diff
@@ -0,0 +1,478 @@
+        {
+            report<PowerOnErrorPSU0PSU1PGOODFail>();
+            std::cerr << "Power_on error : Psu0&PSU1_pgood fail! \n";
+        }
+    }
+}
+
+void MihawkCPLD::analyze()
+{
+    bool psuError = checkPSUdevice();
+    if (psuError)
+    {
+        // If PSU not present, report the error log & event.
+        report<PSUPresentError>();
+        std::cerr << "CPLD : PSU Present Error! \n";
+    }
+}
+
+// Read CPLD_register error code and return the result to analyze.
+int MihawkCPLD::readFromCPLDPSUErrorCode(int bus, int Addr)
+{
+    std::string i2cBus = "/dev/i2c-" + std::to_string(bus);
+
+    // open i2c device(CPLD-PSU-register table)
+    int fd = open(i2cBus.c_str(), O_RDWR | O_CLOEXEC);
+    if (fd < 0)
+    {
+        std::cerr << "Unable to open i2c device \n";
+    }
+
+    // set i2c slave address
+    if (ioctl(fd, I2C_SLAVE_FORCE, Addr) < 0)
+    {
+        std::cerr << "Unable to set device address \n";
+        close(fd);
+    }
+
+    // check whether support i2c function
+    unsigned long funcs = 0;
+    if (ioctl(fd, I2C_FUNCS, &funcs) < 0)
+    {
+        std::cerr << "Not support I2C_FUNCS \n";
+        close(fd);
+    }
+
+    // check whether support i2c-read function
+    if (!(funcs & I2C_FUNC_SMBUS_READ_BYTE_DATA))
+    {
+        std::cerr << "Not support I2C_FUNC_SMBUS_READ_BYTE_DATA \n";
+        close(fd);
+    }
+
+    int statusValue_2;
+    unsigned int statusReg_2 = StatusReg_2;
+
+    statusValue_2 = i2c_smbus_read_byte_data(fd, statusReg_2);
+    close(fd);
+
+    if (statusValue_2 < 0)
+    {
+        statusValue_2 = 0;
+    }
+
+    // return the i2c-read data
+    return statusValue_2;
+}
+
+// Check PSU_present_pin.
+bool MihawkCPLD::checkPSUdevice()
+{
+    bool result = 0;
+    int bin[32], num;
+    int i = 0, length = 0;
+    std::string i2cBus = "/dev/i2c-" + std::to_string(busId);
+
+    // open i2c device(CPLD-PSU-register table)
+    int fd = open(i2cBus.c_str(), O_RDWR | O_CLOEXEC);
+    if (fd < 0)
+    {
+        std::cerr << "Unable to open i2c device \n";
+    }
+
+    // set i2c slave address
+    if (ioctl(fd, I2C_SLAVE_FORCE, slaveAddr) < 0)
+    {
+        std::cerr << "Unable to set device address \n";
+        close(fd);
+    }
+
+    // check whether support i2c function
+    unsigned long funcs = 0;
+    if (ioctl(fd, I2C_FUNCS, &funcs) < 0)
+    {
+        std::cerr << "Not support I2C_FUNCS \n";
+        close(fd);
+    }
+
+    // check whether support i2c-read function
+    if (!(funcs & I2C_FUNC_SMBUS_READ_BYTE_DATA))
+    {
+        std::cerr << "Not support I2C_FUNC_SMBUS_READ_BYTE_DATA \n";
+        close(fd);
+    }
+
+    int statusValue;
+    unsigned int statusReg = StatusReg;
+
+    statusValue = i2c_smbus_read_byte_data(fd, statusReg);
+    close(fd);
+
+    if (statusValue < 0)
+    {
+        std::cerr << "i2c_smbus_read_byte_data failed \n";
+    }
+
+    // Transfer decimal_value to binary_value
+    num = statusValue;
+    do
```

Comment:
The value of statusValue is the value of 1 byte read via I2C. 
This value is the decimal value.

Since the error message of CPLD-register is determined by "bit", each bit represents a different error cause, so this decimal value(statusValue) must be converted to a binary value.

The bin[] array stores the converted binary values. 
For example, bin[0] stores the first bit data, which may be 0 or 1, and bin[1] stores the second bit data. In order, and so on.

The value read in the function of this checkPSUdevice(), via the value of the fourth bit after conversion, represents whether PSU0 is present, 1 means "PSU0 is not present", and 0 means "PSU0 is present".
The value of the fifth bit is PSU1.

Reply:
Is it like this?
So say the read returned a 34.
Does that mean it is 0x22, or 00100010, and so bits 2 and 6 are on?

If that is the case, you can just check for bits being on with mask values and not have to do any looping:
```
if (value & mymask)
```

Even if this is the case, this is something that could just be changed for upstream.

Reply:
Your interpretation is roughly correct but it is a little different.

For example, if the read value is 34, it means 100010.
However, in the field read by the checkPSUdevice() function, the read value is defined as the "**fourth bit is the definition of PSU0**". 
"If the value of this bit is 1, it means PSU0 isn't present"; "if the value of this bit is 0, means PSU0 is present".
The read value is defined as the "**fifth bit is the definition of PSU1**". "If the value of this bit is 1, it means PSU1 isn't present"; "if this bit value is 0, it means PSU1 is present".

Therefore, the fourth and fifth bits of the read value must be checked to determine the "present state" of PSU0 and PSU1.

Reply:
@spinler  Ok, I understand what you mean, I will re-modify this part.